xds/src/main/java/io/grpc/xds/XdsClientImpl.java

```diff
@@ -767,16 +783,18 @@ private void handleRdsResponse(DiscoveryResponse rdsResponse) {
         XdsLogLevel.INFO, "Received RDS response for resources: {0}", routeConfigNames);
 
     // Resolved cluster name for the requested resource, if exists.
-    String clusterName = null;
+    List<EnvoyProtoData.Route> routes = null;
     if (requestedRouteConfig != null) {
-      clusterName = findClusterNameInRouteConfig(requestedRouteConfig, ldsResourceName);
-      if (clusterName == null) {
+      routes = findRoutesInRouteConfig(requestedRouteConfig, ldsResourceName);
+      String errorDetail = validateRoutes(routes);
+      if (errorDetail != null) {
         adsStream.sendNackRequest(
             ADS_TYPE_URL_RDS, ImmutableList.of(adsStream.rdsResourceName),
             rdsResponse.getVersionInfo(),
             "RouteConfiguration " + requestedRouteConfig.getName() + ": cannot find a "
                 + "valid cluster name in any virtual hosts with domains matching: "
-                + ldsResourceName);
+                + ldsResourceName
```

Comment:
Would to rewrite the entire error message instead of just appending an error detail. Also, this entire message becomes the `error_details` field in the message sent to TD, you should avoid having a nested "Error detail:".

Reply:
Rephrased.